File: AIOUSB/samples/USB-DIO-16/standalone_receiver.c
/*
 * This source code looks best with a tab width of 4.
 *
 * All the API functions that DO NOT begin "AIOUSB_" are standard API functions, largely
 * documented in http://accesio.com/MANUALS/USB%20Software%20Reference.pdf. The functions
 * that DO begin with "AIOUSB_" are "extended" API functions added to the Linux
 * implementation. Source code lines in this sample program that are prefixed with the
 * comment "/ * API * /" highlight calls to the AIOUSB API.
 *
 * LIBUSB (http://www.libusb.org/) must be installed on the Linux box (the AIOUSB code
 * was developed using libusb version 1.0.3). After installing libusb, it may also be
 * necessary to set an environment variable so that the libusb and aiousb header files can
 * be located:
 *
 *     export CPATH=/usr/local/include/libusb-1.0/:/usr/local/include/aiousb/
 *
 * Once libusb is installed properly, it should be possible to compile the sample program
 * using the simple command:
 *
 *     make
 *
 * Alternatively, one can "manually" compile the sample program using the command:
 *
 *     g++ receiver.cpp -laiousb -lusb-1.0 -o receiver
 *
 * or, to enable debug features
 *
 *     g++ -ggdb sample.cpp -laiousbdbg -lusb-1.0 -o sample
 */
// }}}


#include <aiousb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

int main( int argc, char **argv ) {
    unsigned long result = AIOUSB_SUCCESS;
    double ReadClockHz = 40000, WriteClockHz = 0;

    result = AIOUSB_Init();
    if ( result < AIOUSB_SUCCESS ) {
        fprintf(stderr,"Error running AIOUSB_Init() ...exiting\n");
        return 2;
    }
    AIOUSB_ListDevices();

    if( argc < 3  || argc > 4 ) {
        printf(\
"\nUSB-DIO-16A test program , AIOUSB version:'%s', date:'%s'.\n\
This program demonstrates a simple high speed receiver program\n\
where the device in question performs digital acquisition using\n\
the DIO_StreamFrame() API call. It uses the device specified by the\n\
HEX_SERIAL_NUMBER that the user specifies.\n\n\
\n\
Usage: %s HEX_SERIAL_NUMBER FRAME_POINTS [CLOCKSPEED]\n\n", AIOUSB_GetVersion(), AIOUSB_GetVersionDate(),argv[0]);
        return 1;
    }
    /* parse command line */
    uint64_t targetSerialNumber = strtoll( *( argv + 1 ), 0, 16 );
    unsigned long framePoints = strtol( *( argv + 2 ), 0, 0 );



    if ( framePoints % 512 != 0 ) {
        fprintf(stderr,"\nFRAME_POINTS (%ld) is NOT an integer number of 512 bytes...", framePoints );
        framePoints = (framePoints / 512) * 512;
        fprintf(stderr,"rounding to %ld\n\n", framePoints );
    }

    if ( argc == 4 ) {
        ReadClockHz = strtol( argv[3] , 0, 10 );
    }


    /*
     * find the device with the specified serial number
     */
    unsigned long deviceIndex = GetDeviceBySerialNumber( targetSerialNumber );
    if ( deviceIndex == diNone ) {
        fprintf(stderr,"No USB-DIO_16 devices were found...exiting\n");
        return 1;
    }

    /*
     * found it; allocate buffer in which to receive streaming DIO data
     */
    unsigned short *const frameData = ( unsigned short * ) malloc( framePoints * sizeof( unsigned short ) );
    if( !frameData  ) {
        fprintf(stderr,"Error; Unable to allocate frame Data buffer...exiting\n");
        exit(3);
    }
    unsigned char outputMask = 0x08;    // receiver: port D is output; C, B and A are input
    unsigned long initialData = 0xffffffff;
    unsigned char tristateMask = 0x00;

    unsigned long transferred;

    /* set up communication parameters */
    AIOUSB_SetCommTimeout( deviceIndex, 16000 );
    AIOUSB_SetStreamingBlockSize( deviceIndex, 512 );

    /* First stop the clocks*/
    {
        double tmpclock = 0;
        result = DIO_StreamSetClocks( deviceIndex, &tmpclock, &tmpclock );
        if ( result != AIOUSB_SUCCESS ) {
            fprintf(stderr,"Can't temporarily stop the clocks: %ld\n", result );
            goto abort;
        }
        result = AIOUSB_ClearFIFO( deviceIndex, CLEAR_FIFO_METHOD_IMMEDIATE_AND_ABORT );
        if ( result != AIOUSB_SUCCESS ) {
            fprintf(stderr,"Can't Clear the FIFO: %ld\n", result );
            goto abort;
        }
    }

    /* configure I/O ports */
    result = DIO_ConfigureEx( deviceIndex, &outputMask, &initialData, &tristateMask );
    if( result < AIOUSB_SUCCESS ) {
        printf( "Error '%s' configuring device at index %lu\n" , AIOUSB_GetResultCodeAsString( result ), deviceIndex );
        DIO_StreamClose( deviceIndex );
        goto abort;
    }

    /* open stream for reading */
    result = DIO_StreamOpen( deviceIndex, AIOUSB_TRUE );
    if( result < AIOUSB_SUCCESS ) {
        fprintf( stderr, "Error '%s' opening read stream for device at index %lu\n" , AIOUSB_GetResultCodeAsString( result ), deviceIndex );
        goto abort;
    }

    /* Set the receiver to  a specific clock frequency */
    result = DIO_StreamSetClocks( deviceIndex, &ReadClockHz, &WriteClockHz );
    if( result < AIOUSB_SUCCESS ) {
        printf( "Error '%s' setting stream clock for device at index %lu\n" , AIOUSB_GetResultCodeAsString( result ), deviceIndex );
        goto abort;
    }

    /* receive frame */
    result = DIO_StreamFrame( deviceIndex, framePoints, frameData, &transferred );
    if( result == AIOUSB_SUCCESS ) {
        if( transferred != framePoints * sizeof( unsigned short ) ) {
            fprintf(stderr,"Received bytes (%ld) was not equal to what was expected (%ld)\n",
                    transferred, framePoints * sizeof( unsigned short ) );
            goto abort;
        }

        printf( "%lu FramePoints successfully read from device at index %lu\n" , framePoints, deviceIndex );
        FILE *fp = fopen("output.dat","w");

        fwrite(frameData,sizeof(frameData[0]),framePoints,fp);

        fclose(fp);


    } else {
        printf("Error running DIO_StreamFrame, result=%ld\n"
               "%lu\n attempted to read %lu bytes, but actually read %lu bytes\n", 
               result,
               deviceIndex, 
               framePoints * sizeof( unsigned short ), 
               transferred 
               );
    }

    DIO_StreamClose( deviceIndex );
    AIOUSB_ClearFIFO( deviceIndex, CLEAR_FIFO_METHOD_IMMEDIATE_AND_ABORT );

    {
        double tmpclock = 0;
        result = DIO_StreamSetClocks( deviceIndex, &tmpclock, &tmpclock );
        if ( result != AIOUSB_SUCCESS ) {
            fprintf(stderr,"Can't temporarily stop the clocks: %ld\n", result );
            goto abort;
        }
    }


 abort:
    free( frameData );

    /* Call before program exits if AIOUSB_Init() succeeded */
    AIOUSB_Exit();

    return ( int ) result;
}

File: AIOUSB/samples/USB-DIO-16/standalone_receiver.py
#!/usr/bin/python

from __future__ import print_function
from AIOUSB import *
import sys
import itertools


result = AIOUSB_SUCCESS;

def main():
    ReadClockHz = 40000
    WriteClockHz = 0;

    result = AIOUSB_Init();
    if ( result < AIOUSB_SUCCESS ):
        fprintf(stderr,"Error running AIOUSB_Init() ...exiting\n");
        exit(2)

    AIOUSB_ListDevices();
    if ( len(sys.argv) < 3  or len(sys.argv) > 4 ):
        print("""
USB-DIO-16A test program , AIOUSB library version %s, %s.
This program demonstrates a simple high speed receiver program
where the device in question performs digital acquisition using
the DIO_StreamFrame() API call. It uses the device specified by the
HEX_SERIAL_NUMBER that the user specifies.

Usage: %s HEX_SERIAL_NUMBER FRAME_POINTS [CLOCKSPEED]
        
Example: %s  0x40e4bc6f1fffffff 512  30000
""" % ( AIOUSB_GetVersion(), AIOUSB_GetVersionDate(),sys.argv[0], sys.argv[0]))
        exit(2)
    
    serialNumber = int(sys.argv[1],16)
    framePoints = int(sys.argv[2])
    if (framePoints / 512) * 512 != framePoints:
        oldfp = framePoints
        framePoints = (framePoints / 512)*512
        print("\nRounding framePoints(%ld) to %ld ( multiple of 512 )\n" % ( oldfp, framePoints) )
                   
    if ( len(sys.argv) == 4 ):
        ReadClockHz = int(sys.argv[3])

    print("Using serialNumber=%x, clock %d\n" % ( serialNumber, ReadClockHz ))
    deviceIndex = GetDeviceBySerialNumber( serialNumber );
    deviceIndex = 0

    result = AIOUSB_SetCommTimeout( deviceIndex, 5000 );

    AIOUSB_SetStreamingBlockSize( deviceIndex, 512 );

    readClock  = new_udp(); udp_assign(readClock,ReadClockHz );
    writeClock = new_udp(); udp_assign(writeClock,0);
    tmpclock  = new_udp(); udp_assign( tmpclock,0 );
    
    # Stop the clocks
    result = DIO_StreamSetClocks( deviceIndex, tmpclock, tmpclock );
    if result != AIOUSB_SUCCESS:
        print("Error '%s' setting stream clocks\n" % ( AIOUSB_GetResultCodeAsString( result ) ))
        sys.exit(1)

    # AIOUSB_ClearFIFO( deviceIndex, CLEAR_FIFO_METHOD_IMMEDIATE_AND_ABORT )


    outputMask = new_ucp(); ucp_assign(outputMask, 0x08 );
    initialData = new_ulp(); ulp_assign( initialData, 0xffffffff );
    tristateMask = new_ucp(); ucp_assign( tristateMask, 0x00 );


    result = DIO_ConfigureEx( deviceIndex, outputMask, initialData, tristateMask );
    if result < AIOUSB_SUCCESS:
        print("Error '%s' configuring device at index %lu\n" % ( AIOUSB_GetResultCodeAsString( result ), deviceIndex ))
        sys.exit(1)

    # open stream for reading
    result = DIO_StreamOpen( deviceIndex, AIOUSB_TRUE );
    if result < AIOUSB_SUCCESS:
        print("Error '%s' opening read stream for device at index %lu\n" % ( AIOUSB_GetResultCodeAsString( result ), deviceIndex ))
        sys.exit(1)


    transferred = new_ulp()
    ulp_assign(transferred, 0)

    # receive frame
    frameData = ushortarray(framePoints)

    # Start clocks
    result = DIO_StreamSetClocks( deviceIndex, readClock, writeClock );
    if result < AIOUSB_SUCCESS:
        print("Error '%s' setting stream clock for device at index %lu\n" % ( AIOUSB_GetResultCodeAsString( result ), deviceIndex ))
        sys.exit(1)

    result = DIO_StreamFrameTmp( deviceIndex, framePoints, frameData, transferred );


    if result == AIOUSB_SUCCESS:
        if ulp_value(transferred) != framePoints * 2:
            print("Received bytes (%ld) was not equal to what was expected (%ld)\n" % (
                    transferred, framePoints * 2 ))
            sys.exit(1)

        print("%lu FramePoints successfully read from device at index %lu\n" % (framePoints, deviceIndex ))

        fp = file("output.dat","w")
        for x in range(0,framePoints):
            fp.write( chr(frameData[x] & 0xff ) )
            fp.write(chr((frameData[x] & 0xff00)>>8))
        fp.close()
    else:
        print("""
Error running DIO_StreamFrame, result=%ld
%lu\n attempted to read %lu bytes, but actually read %lu bytes""" % (
               result,
               deviceIndex, 
               framePoints, 
               ulp_value(transferred)
               ))


    result = DIO_StreamClose( deviceIndex )
    print("Result: %d\n" % result )

    udp_assign( tmpclock,0 );
    result = DIO_StreamSetClocks( deviceIndex, tmpclock, tmpclock );
    AIOUSB_Exit()



if __name__ == '__main__':
  main()




